### PR TITLE
E2E test prototype

### PR DESCRIPTION
**Pull request by norascheuch**
This PR adds a first E2E test that creates a local query, downloads a database, runs the query and waits for the query history item to be read. 

The test can be run locally (documentation provided) as well as on CI .

## Checklist

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

File: .github/workflows/e2e-tests.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+name: Run E2E Playwright tests
+on:
+  push:
```
Comment (norascheuch) [inline .github/workflows/e2e-tests.yml:3]: which actions should we allow/have here? I think it makes sense to allow manual runs and otherwise do it nightly?

File: extensions/ql-vscode/test/e2e/docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
+FROM codercom/code-server:4.20.0
+
+USER root
+
+RUN apt-get update \
+  && apt-get install -y \
+    unzip \
+  && rm -rf /var/lib/apt/lists/*
+
+RUN wget -O /tmp/codeql.zip https://github.com/github/codeql-cli-binaries/releases/download/v2.15.5/codeql-linux64.zip \
```
Comment (norascheuch) [inline extensions/ql-vscode/test/e2e/docker/Dockerfile:10]: we might want to make these silent to not swamp the logs

File: .github/workflows/e2e-tests.yml (hunk 1 above)
Comment (norascheuch) [reply .github/workflows/e2e-tests.yml:3]: I copied from our [CLI test setup](https://github.com/github/vscode-codeql/blob/main/.github/workflows/cli-test.yml#L113) and decided to run them on push to main, merge to main and every night of the week.

File: extensions/ql-vscode/test/e2e/playwright.config.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  },
+
+  /* Configure projects for major browsers */
+  projects: [
+    {
+      name: "chromium",
+      use: { ...devices["Desktop Chrome"] },
+    },
+
+    // {
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/test/e2e/playwright.config.ts:42]: That's a good point! Removed some stuff

File: extensions/ql-vscode/test/e2e/run-query.spec.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  await page.keyboard.press("Enter");
+
+  await page.getByLabel("JavaScript, javascript").locator("a").click({
+    timeout: 60000,
+  });
+
+  // select folder for first query
+  await page
+    .getByText(
+      "Results0 SelectedPress 'Enter' to confirm your input or 'Escape' to cancelOK",
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/test/e2e/run-query.spec.ts:24]: Yeah, this line was taken from the playwright UI locator, it is one of the areas we need to investigate further how to improve.

File: extensions/ql-vscode/test/e2e/README.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
+## VS Code CodeQL E2E Tests
+
+When running the tests locally on a mac a different processor has to be emulated, which makes everythign VERY slow. Therefore we need to add higher timeouts in the test, so that they pass locally.
+
+### How to use locally
+
+Setup
+
+- go to the e2e test folder on your terminal
+- run `docker-compose build`
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/test/e2e/README.md:10]: I would love that!

File: extensions/ql-vscode/test/e2e/docker/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  && unzip /tmp/codeql.zip -d /opt \
```
[the remark this answers is not in the document]
Comment (norascheuch) [reply extensions/ql-vscode/test/e2e/docker/Dockerfile:11]: Thanks!

File: .github/workflows/e2e-tests.yml [hunk 7]
```diff
@@ -40,4 +45,43 @@ jobs:
...
+    env:
+      GH_TOKEN: ${{ github.token }}
+    steps:
+      - name: Create GitHub issue
+        run: |
+          # Set -eu so that we fail if the gh command fails.
+          set -eu
+
+          # Try to find an existing open issue if there is one
+          ISSUE="$(gh issue list --repo "$GITHUB_REPOSITORY" --label "e2e-test-failure" --state "open" --limit 1 --json number -q '.[0].number')"
```
Comment (robertbrignull) [inline .github/workflows/e2e-tests.yml:66]: Note you'll need to create the `e2e-test-failure` label

File: .github/workflows/e2e-tests.yml [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+          if [[ -n "$ISSUE" ]]; then
+            echo "Found open issue number $ISSUE ($GITHUB_SERVER_URL/$GITHUB_REPOSITORY/issues/$ISSUE)"
+          else
+            echo "Did not find an open tracking issue. Creating one."
+
+            ISSUE_BODY="issue-body.md"
+            printf "E2e tests have failed on the default branch.\n\n@github/code-scanning-secexp-reviewers" > "$ISSUE_BODY"
+
+            ISSUE="$(gh issue create --repo "$GITHUB_REPOSITORY" --label "e2e-test-failure" --title "E2@ test failure" --body-file "$ISSUE_BODY")"
```
Comment (robertbrignull) [inline .github/workflows/e2e-tests.yml:76]:
```suggestion
            ISSUE="$(gh issue create --repo "$GITHUB_REPOSITORY" --label "e2e-test-failure" --title "E2E test failure" --body-file "$ISSUE_BODY")"
```
Is this a typo or intentional?

File: .github/workflows/e2e-tests.yml [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            # `gh issue create` returns the full issue URL, not just the number.
+            echo "Created issue with URL $ISSUE"
+          fi
+
+          COMMENT_FILE="comment.md"
+          RUN_URL=$GITHUB_SERVER_URL/$GITHUB_REPOSITORY/actions/runs/$GITHUB_RUN_ID
+          printf 'E2@ test [%s](%s) failed on ref `%s`' "$GITHUB_RUN_ID" "$RUN_URL" "$GITHUB_REF" > "$COMMENT_FILE"
```
Comment (robertbrignull) [inline .github/workflows/e2e-tests.yml:83]:
```suggestion
          printf 'E2E test [%s](%s) failed on ref `%s`' "$GITHUB_RUN_ID" "$RUN_URL" "$GITHUB_REF" > "$COMMENT_FILE"
```
Similar to above, this looks like a mistake

File: .github/workflows/e2e-tests.yml (hunk 8 above)
Comment (robertbrignull) [reply .github/workflows/e2e-tests.yml:76]: No preference on if the second "e" should be capitalised or not. In the PR title it is, but in the issue body text two lines above it isn't. It would be good to be consistent though.

File: .github/workflows/e2e-tests.yml [hunk 10]
```diff
@@ -0,0 +1,87 @@
+name: Run E2E Playwright tests
+on:
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
+  schedule:
```
Comment (charisk) [inline .github/workflows/e2e-tests.yml:7]: Drive-by comment: why do we need to run on schedule if it runs on PRs and merges anyway?

**Review by robertbrignull — APPROVED** [review]
@norascheuch and I just ran through the tests and the docs manually. Everything worked fine, once we got used to the tools.

I'm pro merging this now, even if we wait a bit before adding more tests. Having this merged will show that it works and allow us to gain more experience with it, rather than it just sitting in a branch. The worst case is it becomes flaky and we just disable the tests, so it's not much risk.

File: .github/workflows/e2e-tests.yml (hunk 7 above)
Comment (charisk) [reply .github/workflows/e2e-tests.yml:66]: Need to also make sure that that label exists, otherwise it'll fail

**Comment by charisk** [comment]
> I'm pro merging this now, even if we wait a bit before adding more tests. Having this merged will show that it works and allow us to gain more experience with it, rather than it just sitting in a branch. The worst case is it becomes flaky and we just disable the tests, so it's not much risk.

FWIW I'm happy for us to merge too (although I've not actually reviewed the code) - we discussed with @norascheuch and it seems like getting the tests running will give us some good data about how flaky it might be etc.
